### PR TITLE
[SP-6322] VictorOps updating to new conventions

### PR DESCRIPTION
**Pull request by igorkheif**

### CODE REVIEW

File: steps/victorops/get-oncall/victorops_get_oncall_schedule.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	var respBody map[string]interface{}
+	err = json.Unmarshal(rawRespBody, &respBody)
+	if err != nil {
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
+	}
+
+	// check HTTP error
+	st := base.GetStatus(resp.StatusCode)
```
Comment (hagaishapira) [inline steps/victorops/get-oncall/victorops_get_oncall_schedule.go:57]: Are you sure checking the status here isn't too late?
If you didn't get a success response won't it fail on the unmarshalling before we actually see this is a non successful response?

File: steps/victorops/get-user-data/manifest.yaml [hunk 2]
```diff
@@ -20,3 +20,37 @@ envs:
...
+  - name: first_name
+    description: "First name of the user."
+    type: "string"
+    example:  "Jerry"
+  - name: last_name
+    description: "Last name of the user."
+    type: "string"
+    example:  "Seinfeld"
+  - name: username
+    description: "User's username"
```
Comment (hagaishapira) [inline steps/victorops/get-user-data/manifest.yaml:34]: all descriptions should have dots at the end of the sentences.

File: steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go [hunk 3]
```diff
@@ -25,91 +26,74 @@ type ackRequest struct {
...
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
 	}
 
-	//read response
+	// read response
 	defer resp.Body.Close()
 	rawRespBody, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return step.ExitCodeFailure, nil, err
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
```
Comment (hagaishapira) [inline steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go:65]: @oriser are we sure we want the same generic message for all these failure cases?

File: steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go [hunk 4]
```diff
@@ -25,91 +26,74 @@ type ackRequest struct {
...
+	if err != nil {
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
 	}
 
-	var responseBody ackResponse
-	err = json.Unmarshal(rawRespBody, &responseBody)
-	if err != nil {
-		return step.ExitCodeFailure, nil, err
+	// check HTTP error
+	st := base.GetStatus(resp.StatusCode)
```
Comment (hagaishapira) [inline steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go:75]: Again with the place of when to check the status

File: steps/victorops/incident/create-note/manifest.yaml [hunk 5]
```diff
@@ -33,6 +33,10 @@ outputs:
   - name: name
     description: "The unique name of your note."
     type: "string"
+  - name: success
```
Comment (hagaishapira) [inline steps/victorops/incident/create-note/manifest.yaml:36]: This is actually not needed as we have the step status which if it is success then the note should be created

File: steps/victorops/incident/create-note/victorops_incident_create_note.go [hunk 6]
```diff
@@ -28,80 +31,70 @@ type createRequestMessage struct {
...
 	request.Header.Set("X-VO-Api-Key", s.ApiKey)
 	request.Header.Set("Content-Type", "application/json")
 
-	//send request
+	// send request
 	client := http.Client{}
 	resp, err := client.Do(request)
 	if err != nil {
-		return step.ExitCodeFailure, nil, err
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
```
Comment (hagaishapira) [inline steps/victorops/incident/create-note/victorops_incident_create_note.go:66]: same question about the generic message

**Review by hagaishapira — CHANGES_REQUESTED** [review]

File: steps/victorops/base/base.go [hunk 7]
```diff
@@ -1,6 +1,69 @@
...
+		retStatus = status.OK
+	case 400:
+		retStatus = status.Status{
+			Code:    codes.InvalidArgument,
+			Message: "problem with the request arguments",
+			Verbose: "",
+		}
+	case 401:
+		retStatus = status.Status{
+			Code:    codes.InvalidArgument,
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/victorops/base/base.go:33]: In Victorops lingo 401 is "Authentication parameters missing". I can see it as either. Up to you.

File: steps/victorops/base/base.go [hunk 8]
```diff
@@ -1,6 +1,69 @@
...
+			Message: "authentication parameters missing",
+			Verbose: "",
+		}
+	case 403:
+		// TODO: should this still be InvalidArgument?
+		retStatus = status.Status{
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/victorops/base/base.go:39]: This is "Authentication failed or rate-limit reached"

File: steps/victorops/get-oncall/victorops_get_oncall_schedule.go (hunk 1 above)
Comment (igorkheif) [reply steps/victorops/get-oncall/victorops_get_oncall_schedule.go:57]: Let's f2f about this

File: steps/victorops/base/base.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (oriser) [reply steps/victorops/base/base.go:33]: sounds legit

File: steps/victorops/base/go.mod [hunk 9]
```diff
@@ -1,3 +1,7 @@
 module github.com/stackpulse/steps/opsgenie/base
 
 go 1.14
```
Comment (oriser) [inline steps/victorops/base/go.mod:3]:
```suggestion
go 1.16
```

File: steps/victorops/get-oncall/victorops_get_oncall_schedule_test.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+										"onCalluser": map[string]interface{}{
+											"username": "spvictor",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
```
Comment (oriser) [inline steps/victorops/get-oncall/victorops_get_oncall_schedule_test.go:70]: Why not using the matcher interfaces?
It's very hard to read it that way

File: steps/victorops/get-user-data/victorops_get_user_data.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+type VictorOpsGetUserData struct {
+	args Args
+}
+
+func (v *VictorOpsGetUserData) Init() step.Status {
+	return env.Parse(&v.args)
+}
+
+func (v *VictorOpsGetUserData) Run() (step.Output, step.Status) {
+	victoropsClient := victorops.NewClient(v.args.ApiID, v.args.ApiKey, "https://api.victorops.com")
```
Comment (oriser) [inline steps/victorops/get-user-data/victorops_get_user_data.go:29]: Why are you using the VO library here but in the on call step you are using HTTP request?

File: steps/victorops/get-user-data/victorops_get_user_data.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		"verified":              user.Verified,
+		"api_object":            user,
+
+		// BC
+		"firstName":           user.FirstName,
+		"lastName":            user.LastName,
+		"createdAt":           user.CreatedAt,
+		"passwordLastUpdated": user.PasswordLastUpdated,
+		"Admin":               user.Admin,
+		"ExpirationHours":     user.ExpirationHours,
```
Comment (oriser) [inline steps/victorops/get-user-data/victorops_get_user_data.go:65]: Did we check if someone using this step? If no one is using, we can break it.
If you don't sure how to check DM me and let's do it together

File: steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go (hunk 3 above)
Comment (oriser) [reply steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go:65]: @igorkheif why did you decided to return a generic message here?

File: steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go [hunk 13]
```diff
@@ -25,91 +26,74 @@ type ackRequest struct {
...
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
 	}
 
-	//check HTTP error
-	if resp.StatusCode != http.StatusOK {
-		return step.ExitCodeFailure, nil, fmt.Errorf("failed to acknowledge incident. got response code: %d: %s", resp.StatusCode, string(rawRespBody))
+	var respBody map[string]interface{}
+	err = json.Unmarshal(rawRespBody, &respBody)
+	if err != nil {
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
```
Comment (oriser) [inline steps/victorops/incident/acknowledge/victorops_incident_acknowledge.go:71]: Here for example, I think the most useful status is to write in the message the error returned from the unmarshaling and in the verbose, display the original respones (`fmt.Sprintf("original response: %s", string(rawRespBody))`)

File: steps/victorops/incident/create-note/victorops_incident_create_note.go [hunk 14]
```diff
@@ -28,80 +31,70 @@ type createRequestMessage struct {
...
-	}
-
-	var responseBody map[string]interface{}
-	err = json.Unmarshal(rawRespBody, &responseBody)
+	// unmarshal response
+	var respBody map[string]interface{}
+	err = json.Unmarshal(rawRespBody, &respBody)
 	if err != nil {
-		return step.ExitCodeFailure, nil, err
+		return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
```
Comment (oriser) [inline steps/victorops/incident/create-note/victorops_incident_create_note.go:80]: Same here

File: steps/victorops/incident/get/victorops_get_incident.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+		return nil, status.FromError(codes.InvalidArgument, "could not get incident", err)
+	}
+
+	retStatus := base.GetStatus(details.StatusCode)
+	if retStatus.Failed() {
+		// get raw response body object
+		var respBody map[string]interface{}
+		err = json.Unmarshal([]byte(details.ResponseBody), &respBody)
+		if err != nil {
+			return nil, status.FromError(codes.Internal, base.GenericFailMessage, err)
```
Comment (oriser) [inline steps/victorops/incident/get/victorops_get_incident.go:38]: same

File: steps/victorops/incident/create-note/victorops_incident_create_note_test.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	req.Header.Set("X-VO-Api-Id", os.Getenv("VICTOROPS_API_ID"))
+	req.Header.Set("X-VO-Api-Key", os.Getenv("VICTOROPS_API_KEY"))
+
+	// Fetch Request
+	resp, err := client.Do(req)
```
Comment (oriser) [inline steps/victorops/incident/create-note/victorops_incident_create_note_test.go:43]: You can export it to the base

File: steps/victorops/incident/get/victorops_get_incident.go [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+	base.Args
+	IncidentNumber int `env:"INCIDENT_NUMBER,required"`
+}
+
+func (s *IncidentGet) Init() status.Status {
+	return env.Parse(s)
+}
+
+func (s *IncidentGet) Run() (step.Output, status.Status) {
+	victoropsClient := victorops.NewClient(s.ApiID, s.ApiKey, "https://api.victorops.com")
```
Comment (oriser) [inline steps/victorops/incident/get/victorops_get_incident.go:25]: You can create the VO client in the base

**Review by oriser — COMMENTED** [review]
See comments

File: steps/victorops/base/base.go [hunk 18]
```diff
@@ -1,6 +1,69 @@
 package base
 
+import (
+	"fmt"
+
+	"github.com/stackpulse/steps-sdk-go/step"
+
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/victorops/base/base.go:7]: 👍

File: steps/victorops/base/go.mod (hunk 9 above)
Comment (igorkheif) [reply steps/victorops/base/go.mod:3]: 👍

File: steps/victorops/get-user-data/manifest.yaml (hunk 2 above)
Comment (igorkheif) [reply steps/victorops/get-user-data/manifest.yaml:34]: 👍

**Pull request closed without merging**